File: python/array.py
from typing import TypeVar, Generic, List, Optional, Callable

T = TypeVar('T')

class FixedArray(Generic[T]):
    def __init__(self, capacity: int):
        if capacity <= 0:
            raise ValueError("Capacity must be positive")
        self._capacity: int = capacity
        self._length: int = 0
        self._data: List[Optional[T]] = [None] * capacity

    @property
    def capacity(self) -> int:
        return self._capacity

    @property
    def length(self) -> int:
        return self._length

    @property
    def is_empty(self) -> bool:
        return self._length == 0

    @property
    def is_full(self) -> bool:
        return self._length == self._capacity

    def at(self, index: int) -> Optional[T]:
        if index < 0 or index >= self._length:
            raise IndexError("Index out of bounds")
        return self._data[index]

    def set(self, index: int, value: T) -> None:
        if index < 0 or index >= self._length:
            raise IndexError("Index out of bounds")
        self._data[index] = value

    def push(self, value: T) -> None:
        if self.is_full:
            raise RuntimeError("Array is full")
        self._data[self._length] = value
        self._length += 1

    def pop(self) -> Optional[T]:
        if self.is_empty:
            raise RuntimeError("Array is empty")
        self._length -= 1
        value = self._data[self._length]
        self._data[self._length] = None
        return value

    def insert(self, index: int, value: T) -> None:
        if self.is_full:
            raise RuntimeError("Array is full")
        if index < 0 or index > self._length:
            raise IndexError("Index out of bounds")
        for i in range(self._length, index, -1):
            self._data[i] = self._data[i - 1]
        self._data[index] = value
        self._length += 1

    def remove(self, index: int) -> Optional[T]:
        if self.is_empty:
            raise RuntimeError("Array is empty")
        if index < 0 or index >= self._length:
            raise IndexError("Index out of bounds")
        value = self._data[index]
        for i in range(index, self._length - 1):
            self._data[i] = self._data[i + 1]
        self._length -= 1
        self._data[self._length] = None
        return value

    def find(self, value: T, equal: Callable[[T, T], bool] = lambda x, y: x == y) -> int:
        for i in range(self._length):
            if self._data[i] is not None and equal(self._data[i], value):
                return i
        return -1

    def to_list(self) -> List[Optional[T]]:
        return self._data[:self._length]

    def clear(self) -> None:
        self._data = [None] * self._capacity
        self._length = 0

    def __getitem__(self, index: int) -> Optional[T]:
        return self.at(index)

    def __setitem__(self, index: int, value: T) -> None:
        self.set(index, value)

    def __len__(self) -> int:
        return self._length

    def __iter__(self):
        for i in range(self._length):
            yield self._data[i]

    def __str__(self) -> str:
        return str(self.to_list())


class DynamicArray(Generic[T]):
    def __init__(self, initial_capacity: int = 10):
        if initial_capacity <= 0:
            raise ValueError("Initial capacity must be positive")
        self._capacity: int = initial_capacity
        self._length: int = 0
        self._data: List[Optional[T]] = [None] * initial_capacity
        self._growth_factor: int = 2

    @property
    def capacity(self) -> int:
        return self._capacity

    @property
    def length(self) -> int:
        return self._length

    @property
    def is_empty(self) -> bool:
        return self._length == 0

    def _resize(self) -> None:
        new_capacity = self._capacity * self._growth_factor
        new_data: List[Optional[T]] = [None] * new_capacity
        for i in range(self._length):
            new_data[i] = self._data[i]
        self._data = new_data
        self._capacity = new_capacity

    def at(self, index: int) -> Optional[T]:
        if index < 0 or index >= self._length:
            raise IndexError("Index out of bounds")
        return self._data[index]

    def set(self, index: int, value: T) -> None:
        if index < 0 or index >= self._length:
            raise IndexError("Index out of bounds")
        self._data[index] = value

    def push(self, value: T) -> None:
        if self._length >= self._capacity:
            self._resize()
        self._data[self._length] = value
        self._length += 1

    def pop(self) -> Optional[T]:
        if self.is_empty:
            raise RuntimeError("Array is empty")
        self._length -= 1
        value = self._data[self._length]
        self._data[self._length] = None
        return value

    def insert(self, index: int, value: T) -> None:
        if index < 0 or index > self._length:
            raise IndexError("Index out of bounds")
        if self._length >= self._capacity:
            self._resize()
        for i in range(self._length, index, -1):
            self._data[i] = self._data[i - 1]
        self._data[index] = value
        self._length += 1

    def remove(self, index: int) -> Optional[T]:
        if self.is_empty:
            raise RuntimeError("Array is empty")
        if index < 0 or index >= self._length:
            raise IndexError("Index out of bounds")
        value = self._data[index]
        for i in range(index, self._length - 1):
            self._data[i] = self._data[i + 1]
        self._length -= 1
        self._data[self._length] = None
        return value

    def find(self, value: T, equal: Callable[[T, T], bool] = lambda x, y: x == y) -> int:
        for i in range(self._length):
            if self._data[i] is not None and equal(self._data[i], value):
                return i
        return -1

    def to_list(self) -> List[Optional[T]]:
        return self._data[:self._length]

    def clear(self) -> None:
        self._data = [None] * 10
        self._capacity = 10
        self._length = 0

    def sort(self, key: Optional[Callable[[T], any]] = None, reverse: bool = False) -> None:
        arr = [x for x in self._data[:self._length] if x is not None]
        arr.sort(key=key, reverse=reverse)
        for i in range(len(arr)):
            self._data[i] = arr[i]
        for i in range(len(arr), self._length):
            self._data[i] = None

    def __getitem__(self, index: int) -> Optional[T]:
        return self.at(index)

    def __setitem__(self, index: int, value: T) -> None:
        self.set(index, value)

    def __len__(self) -> int:
        return self._length

    def __iter__(self):
        for i in range(self._length):
            yield self._data[i]

    def __str__(self) -> str:
        return str(self.to_list())


if __name__ == "__main__":
    print("FixedArray Example:")
    fa = FixedArray[int](5)
    fa.push(1)
    fa.push(2)
    fa.push(3)
    print(f"FixedArray: {fa}")
    print("Iterating with for loop:")
    for val in fa:
        print(val, end=' ')
    print()
    
    print("\nDynamicArray Example:")
    da = DynamicArray[int](2)
    da.push(10)
    da.push(20)
    da.push(30)
    print(f"DynamicArray: {da}")
    print(f"Capacity: {da.capacity}")
    print("Iterating with for loop:")
    for val in da:
        print(val, end=' ')
    print()

File: go/array.go
package main

import (
	"errors"
)

type FixedArray[T any] struct {
	capacity int
	length   int
	data     []T
}

func NewFixedArray[T any](capacity int) (*FixedArray[T], error) {
	if capacity <= 0 {
		return nil, errors.New("capacity must be positive")
	}
	return &FixedArray[T]{
		capacity: capacity,
		length:   0,
		data:     make([]T, capacity),
	}, nil
}

func (fa *FixedArray[T]) Capacity() int {
	return fa.capacity
}

func (fa *FixedArray[T]) Length() int {
	return fa.length
}

func (fa *FixedArray[T]) IsEmpty() bool {
	return fa.length == 0
}

func (fa *FixedArray[T]) IsFull() bool {
	return fa.length == fa.capacity
}

func (fa *FixedArray[T]) At(index int) (T, error) {
	var zero T
	if index < 0 || index >= fa.length {
		return zero, errors.New("index out of bounds")
	}
	return fa.data[index], nil
}

func (fa *FixedArray[T]) Set(index int, value T) error {
	if index < 0 || index >= fa.length {
		return errors.New("index out of bounds")
	}
	fa.data[index] = value
	return nil
}

func (fa *FixedArray[T]) Push(value T) error {
	if fa.IsFull() {
		return errors.New("array is full")
	}
	fa.data[fa.length] = value
	fa.length++
	return nil
}

func (fa *FixedArray[T]) Pop() (T, error) {
	var zero T
	if fa.IsEmpty() {
		return zero, errors.New("array is empty")
	}
	fa.length--
	value := fa.data[fa.length]
	var zeroVal T
	fa.data[fa.length] = zeroVal
	return value, nil
}

func (fa *FixedArray[T]) Insert(index int, value T) error {
	if fa.IsFull() {
		return errors.New("array is full")
	}
	if index < 0 || index > fa.length {
		return errors.New("index out of bounds")
	}
	for i := fa.length; i > index; i-- {
		fa.data[i] = fa.data[i-1]
	}
	fa.data[index] = value
	fa.length++
	return nil
}

func (fa *FixedArray[T]) Remove(index int) (T, error) {
	var zero T
	if fa.IsEmpty() {
		return zero, errors.New("array is empty")
	}
	if index < 0 || index >= fa.length {
		return zero, errors.New("index out of bounds")
	}
	value := fa.data[index]
	for i := index; i < fa.length-1; i++ {
		fa.data[i] = fa.data[i+1]
	}
	fa.length--
	var zeroVal T
	fa.data[fa.length] = zeroVal
	return value, nil
}

func (fa *FixedArray[T]) Find(value T, equal func(a, b T) bool) int {
	for i := 0; i < fa.length; i++ {
		if equal(fa.data[i], value) {
			return i
		}
	}
	return -1
}

func (fa *FixedArray[T]) ToArray() []T {
	result := make([]T, fa.length)
	copy(result, fa.data[:fa.length])
	return result
}

func (fa *FixedArray[T]) Clear() {
	fa.data = make([]T, fa.capacity)
	fa.length = 0
}

func (fa *FixedArray[T]) Iter() func() (T, bool) {
	index := 0
	length := fa.length
	return func() (T, bool) {
		if index >= length {
			var zero T
			return zero, false
		}
		value := fa.data[index]
		index++
		return value, true
	}
}

type DynamicArray[T any] struct {
	capacity     int
	length       int
	data         []T
	growthFactor int
}

func NewDynamicArray[T any](initialCapacity int) (*DynamicArray[T], error) {
	if initialCapacity <= 0 {
		return nil, errors.New("initial capacity must be positive")
	}
	return &DynamicArray[T]{
		capacity:     initialCapacity,
		length:       0,
		data:         make([]T, initialCapacity),
		growthFactor: 2,
	}, nil
}

func (da *DynamicArray[T]) Capacity() int {
	return da.capacity
}

func (da *DynamicArray[T]) Length() int {
	return da.length
}

func (da *DynamicArray[T]) IsEmpty() bool {
	return da.length == 0
}

func (da *DynamicArray[T]) resize() {
	newCapacity := da.capacity * da.growthFactor
	newData := make([]T, newCapacity)
	copy(newData, da.data)
	da.data = newData
	da.capacity = newCapacity
}

func (da *DynamicArray[T]) At(index int) (T, error) {
	var zero T
	if index < 0 || index >= da.length {
		return zero, errors.New("index out of bounds")
	}
	return da.data[index], nil
}

func (da *DynamicArray[T]) Set(index int, value T) error {
	if index < 0 || index >= da.length {
		return errors.New("index out of bounds")
	}
	da.data[index] = value
	return nil
}

func (da *DynamicArray[T]) Push(value T) {
	if da.length >= da.capacity {
		da.resize()
	}
	da.data[da.length] = value
	da.length++
}

func (da *DynamicArray[T]) Pop() (T, error) {
	var zero T
	if da.IsEmpty() {
		return zero, errors.New("array is empty")
	}
	da.length--
	value := da.data[da.length]
	var zeroVal T
	da.data[da.length] = zeroVal
	return value, nil
}

func (da *DynamicArray[T]) Insert(index int, value T) error {
	if index < 0 || index > da.length {
		return errors.New("index out of bounds")
	}
	if da.length >= da.capacity {
		da.resize()
	}
	for i := da.length; i > index; i-- {
		da.data[i] = da.data[i-1]
	}
	da.data[index] = value
	da.length++
	return nil
}

func (da *DynamicArray[T]) Remove(index int) (T, error) {
	var zero T
	if da.IsEmpty() {
		return zero, errors.New("array is empty")
	}
	if index < 0 || index >= da.length {
		return zero, errors.New("index out of bounds")
	}
	value := da.data[index]
	for i := index; i < da.length-1; i++ {
		da.data[i] = da.data[i+1]
	}
	da.length--
	var zeroVal T
	da.data[da.length] = zeroVal
	return value, nil
}

func (da *DynamicArray[T]) Find(value T, equal func(a, b T) bool) int {
	for i := 0; i < da.length; i++ {
		if equal(da.data[i], value) {
			return i
		}
	}
	return -1
}

func (da *DynamicArray[T]) ToArray() []T {
	result := make([]T, da.length)
	copy(result, da.data[:da.length])
	return result
}

func (da *DynamicArray[T]) Clear() {
	da.data = make([]T, 10)
	da.capacity = 10
	da.length = 0
}

func (da *DynamicArray[T]) Sort(less func(a, b T) bool) {
	for i := 0; i < da.length; i++ {
		for j := i + 1; j < da.length; j++ {
			if less(da.data[j], da.data[i]) {
				da.data[i], da.data[j] = da.data[j], da.data[i]
			}
		}
	}
}

func (da *DynamicArray[T]) Iter() func() (T, bool) {
	index := 0
	length := da.length
	return func() (T, bool) {
		if index >= length {
			var zero T
			return zero, false
		}
		value := da.data[index]
		index++
		return value, true
	}
}
